### PR TITLE
Import tracking

### PR DESCRIPTION
**Pull request by jkanem**
The serialization of the workflow, which results in dispatch_source.py, now includes imports from files external to the Covalent code-base.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation, VERSION, and CHANGELOG accordingly.
- [x] I have read the CONTRIBUTING document.

### CODE REVIEW

**Review by wjcunningham7 — COMMENTED** [review]
I'm getting a problem in a very simple example, written in the Python console, which is not inspectable:
```
import covalent as ct
from covalent._shared_files.utils import get_serialized_function_str
@ct.electron
def test_func(input1: str, input2: int) -> str:
    return f"{input1} {input2}"
funcstr = get_serialized_function_str(test_func)
print(funcstr)
```

This says the source is not inspectable.

**Comment by wjcunningham7** [comment]
> [CLA assistant check](https://cla-assistant.io/AgnostiqHQ/covalent?pullRequest=66) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/AgnostiqHQ/covalent?pullRequest=66) before we can accept your contribution.
> 
> **Jalani Kanem** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/AgnostiqHQ/covalent?pullRequest=66) it.

@jkanem make sure all your commits use the email associated with your GitHub account.  You can do this with `git commit --amend --author="My Name <my@email.com>"` followed by a force push to your branch.

File: covalent/_results_manager/result.py
Comment (jkanem) [reply covalent/_results_manager/result.py]: Done.

**Comment by jkanem** [comment]
> I'm getting a problem in a very simple example, written in the Python console, which is not inspectable:
> 
> ```
> import covalent as ct
> from covalent._shared_files.utils import get_serialized_function_str
> @ct.electron
> def test_func(input1: str, input2: int) -> str:
>     return f"{input1} {input2}"
> funcstr = get_serialized_function_str(test_func)
> print(funcstr)
> ```
> 
> This says the source is not inspectable.

The main work-horse of this functionality is the inspect module. Which works with files. Trying to use it in a console or Jupyter results in `OSError: could not get source code`.

File: covalent/_shared_files/utils.py
Comment (kessler-frost) [inline covalent/_shared_files/utils.py:264]: And this be,
```python
if node.module.split(".")[0] in [
          "covalent",
          "covalent_dispatcher",
      ]:
          cova_module = True
```

File: covalent/_shared_files/utils.py
Comment (kessler-frost) [inline covalent/_shared_files/utils.py:274]: Similar to above,

```python
if subnode.name.split(".")[0] in [
        "covalent",
        "covalent_dispatcher",
    ]:
        cova_module = True
```

File: covalent/executor/base.py
Comment (kessler-frost) [inline covalent/executor/base.py:320]: This seems incomplete.

**Review by kessler-frost — COMMENTED** [review]
For the following workflow ran on a jupyter notebook,

```python
import covalent as ct

@ct.electron
def sample_task(x):
    """Sample task."""

    return x ** 2

@ct.lattice
def sample_workflow(a, b):
    """Sample workflow."""

    c = sample_task(a)
    d = sample_task(b)
    return c, d
```
Sadly the `dispatch_source` file that I'm getting contains a huge list of imports. Maybe there is an already open issue for that or maybe we want to solve it in this one, I'm not sure.

File: covalent/_results_manager/result.py
Comment (kessler-frost) [inline covalent/_results_manager/result.py:515]: Also this as `if dupes`.

Comment (jkanem) [reply covalent/_results_manager/result.py:515]: I prefer `len(dupes)` because it lets the reader quickly know that dupes is a container which should not be empty. As opposed to a bool or some other object that could return True/False when asked `if dupes`

Comment (kessler-frost) [reply covalent/_results_manager/result.py:515]: Ah, I see your point.

File: covalent/executor/base.py
Comment (jkanem) [reply covalent/executor/base.py]: Fixed.

File: covalent/_shared_files/utils.py
Comment (jkanem) [reply covalent/_shared_files/utils.py]: Fixed

File: covalent/_shared_files/utils.py
Comment (jkanem) [reply covalent/_shared_files/utils.py]: Fixed

File: covalent/_shared_files/utils.py
Comment (jkanem) [reply covalent/_shared_files/utils.py]: I don't know. I think the former is more readable. But I don't care that much, so I changed it.

**Comment by jkanem** [comment]
> For the following workflow ran on a jupyter notebook,
> 
> ```python
> import covalent as ct
> 
> @ct.electron
> def sample_task(x):
>     """Sample task."""
> 
>     return x ** 2
> 
> @ct.lattice
> def sample_workflow(a, b):
>     """Sample workflow."""
> 
>     c = sample_task(a)
>     d = sample_task(b)
>     return c, d
> ```
> 
> Sadly the `dispatch_source` file that I'm getting contains a huge list of imports. Maybe there is an already open issue for that or maybe we want to solve it in this one, I'm not sure.
This is a Jupyter issue; Jupyter is external to the covalent code-base, and is seen as part of the end-user scripts. Possibly, making this work with Jupyter should be a separate issue.

**Comment by kessler-frost** [comment]
I agree. Just curious, were you able to reproduce this too?

**Review by kessler-frost — APPROVED** [review]

**Comment by kessler-frost** [comment]
@jkanem Approved! Looks great, I'll merge this to develop once the tests pass and its synced with develop.

**Comment by jkanem** [comment]
I think this is ready for another look. The scanning for imports is now behind a configuration parameter in the `sdk` section:
```
[sdk]
full_dispatch_source = "false"
```
I'm open to a different name or section.
As mentioned, this will cause a regression in that, by default, covalent-related decorators will not be commented out in dispatch_source.py.

**Comment by jkanem** [comment]
This shouldn't be merged until https://github.com/AgnostiqHQ/covalent/pull/92 is.

**Comment by kessler-frost** [comment]
@jkanem Do you think this PR is also blocked by #80 ?

**Comment by jkanem** [comment]
> @jkanem Do you think this PR is also blocked by #80 ?

No, I think this is a (non-ideal) solution to https://github.com/AgnostiqHQ/covalent/issues/80. This gets rid of the unreasonable time that it takes to get function strings, but it also gets rid of the import statements in dispatch_source.py.

**Comment by jkanem** [comment]
This PR is now deprecated  by https://github.com/AgnostiqHQ/covalent/pull/123

**Pull request closed without merging**
